tracer/src/Datadog.Trace/TracerManagerFactory.cs

```diff
@@ -251,32 +250,75 @@ protected virtual bool ShouldEnableRemoteConfiguration(ImmutableTracerSettings s
 
         protected virtual ITraceSampler GetSampler(ImmutableTracerSettings settings)
         {
+            // ISamplingRule is used to implement, in order of precedence:
+            // - custom sampling rules
+            //   - remote custom rules (provenance: "customer")
+            //   - remote dynamic rules (provenance: "dynamic")
+            //   - local custom rules (provenance: "local"/none) = DD_TRACE_SAMPLING_RULES
+            // - global sampling rate
+            //   - remote
+            //   - local = DD_TRACE_SAMPLE_RATE
+            // - agent sampling rates (as a single rule)
+
+            // Note: the order that rules are registered is important, as they are evaluated in order.
+            // The first rule that matches will be used to determine the sampling rate.
+
+            // Remote and local "custom" sampling rules:
+            // Unlike most settings, local custom sampling rules configuration (DD_TRACE_SAMPLING_RULES) is not
+            // simply overriden with the remote configuration. Instead, remote rules are merged with local rules,
+            // with remote rules taking precedence.
+
             var sampler = new TraceSampler(new TracerRateLimiter(settings.MaxTracesSubmittedPerSecondInternal));
-            var samplingRules = settings.CustomSamplingRulesInternal;
-            var patternFormatIsValid = SamplingRulesFormat.IsValid(settings.CustomSamplingRulesFormat, out var samplingRulesFormat);
 
-            if (patternFormatIsValid && !string.IsNullOrWhiteSpace(samplingRules))
+            // remote sampling rules
+            var remoteSamplingRulesJson = settings.RemoteSamplingRules;
+
+            if (!string.IsNullOrWhiteSpace(remoteSamplingRulesJson))
             {
-                foreach (var rule in CustomSamplingRule.BuildFromConfigurationString(samplingRules, samplingRulesFormat, RegexBuilder.DefaultTimeout))
+                var remoteSamplingRules =
+                    RemoteCustomSamplingRule.BuildFromConfigurationString(remoteSamplingRulesJson, RegexBuilder.DefaultTimeout);
+
+                foreach (var remoteSamplingRule in remoteSamplingRules)
                 {
-                    sampler.RegisterRule(rule);
+                    sampler.RegisterRule(remoteSamplingRule);
```

Comment:
It looks like the various remote rules are just added in order. Do we need to order them ourselves, such that remote custom rules are inserted before remote dynamic rules? Or are they sent to the tracer in the correct order?

Reply:
Good question. The remote rules are sent via RC in order, so we don't need to to reorder them. I initially had code to add the `provenance: "customer"` rules first and then `provenance: "dynamic"`, just to be sure. I then realized this would not be future-proof is other `provenance` values are added later, so we'll just rely on the order that comes from RC.